### PR TITLE
feat(ConnectionManager): enhance connection manager

### PR DESCRIPTION
**Pull request by juliangieseke**
⚠️ Depends on https://github.com/dcos-labs/connections/pull/3 ⚠️

---

⚠️ BREAKING CHANGE⚠️ 

---

This commit completely rewrites the Connection Manager.

1. It changes the interface from `queue`/`dequeue` to a single `schedule` method
3. It uses one loop to process the list instead of fiddling around with different events…

---

I have no idea how to write proper tests for this 🤔

### CODE REVIEW

File: src/ConnectionManager.js [hunk 1]
```diff
@@ -1,19 +1,15 @@
...
- * waiting line.
- */
+import ConnectionQueueItem from "./ConnectionQueueItem";
+
 export default class ConnectionManager {
   /**
    * Initializes an Instance of ConnectionManager
    *
    * @param {int} maxConnections – max open connections
+   * @param {float} threshold - only kill connections with a priority < newPriority * threshold
```
Comment (juliangieseke) [inline src/ConnectionManager.js]: wrong doc

File: src/ConnectionManager.js [hunk 2]
```diff
@@ -1,19 +1,15 @@
...
    */
-  constructor(maxConnections = 6) {
+  constructor(maxConnections = 6, maxAge = 6000, maxAutoPriority = 3) {
```
Comment (juliangieseke) [inline src/ConnectionManager.js]: change to options object

File: src/ConnectionManager.js [hunk 3]
```diff
@@ -28,141 +24,162 @@ export default class ConnectionManager {
...
-        connection,
-        priority
-      );
-    }
+        // merge them again.
+        context.list = openList.concat(waitingList);
+
+        // still waiting items? 
+        if(waitingList.size > 0) {
+          const delay = (maxConnections * openList.size);
```
Comment (juliangieseke) [inline src/ConnectionManager.js]: I tried to make this more "smart", but in the end it only increases complexity and doesnt realy brings a benefit…

File: src/ConnectionManager.js [hunk 4]
```diff
@@ -28,141 +24,162 @@ export default class ConnectionManager {
...
 
-    connection.addListener(ConnectionEvent.ABORT, this.handleConnectionAbort);
+        // create a new QueueItem to have the correct default priority
+        const item = new ConnectionQueueItem(connection, priority);
 
-    connection.addListener(
-      ConnectionEvent.COMPLETE,
-      this.handleConnectionComplete
-    );
+        // if we got a (now) closed connection, nothing to do.
```
Comment (juliangieseke) [inline src/ConnectionManager.js]: wrong comment

File: src/ConnectionManager.js [hunk 5]
```diff
@@ -28,141 +24,162 @@ export default class ConnectionManager {
...
+        if (connection.state === AbstractConnection.CLOSED) {
+          return;
+        }
+    
+        // add, sort and process
```
Comment (juliangieseke) [inline src/ConnectionManager.js]: wrong comment

File: src/index.js [hunk 6]
```diff
@@ -1,3 +1,6 @@
-import { default as ConnectionManager } from "./ConnectionManager";
+import { default as CM } from "./ConnectionManager";
+export { default as ConnectionQueueItem } from "./ConnectionQueueItem";
 
-export default new ConnectionManager();
+export default new CM();
```
Comment (juliangieseke) [inline src/index.js]: can we get rid of this asap, please? :)

**Comment by nLight** [comment]
Please adjust commit message for the breaking change according to https://gist.github.com/stephenparish/9941e89d80e2bc58a153#breaking-changes (Before, After thingie)

**Comment by juliangieseke** [comment]
done 👍

File: src/ConnectionManager.js [hunk 7]
```diff
@@ -28,141 +23,125 @@ export default class ConnectionManager {
...
+        // store all open items
+        let openList = context.list.filter(
+          listItem => listItem.connection.state === AbstractConnection.OPEN
+        );
+
+        let waitingList = context.list
+          .filter(
+            listItem => listItem.connection.state === AbstractConnection.INIT
+          )
+          // this code would increase priority on old items, after they lived in the line for more then 30s.
```
Comment (DanielMSchmidt) [inline src/ConnectionManager.js]: Is there a good reason this should stay?

Comment (juliangieseke) [reply src/ConnectionManager.js]: we have to decide if we want that behaviour or not, thats the reason why i kept it as a comment

**Pull request closed without merging**